### PR TITLE
sns, stream: enable sns filter policies;

### PR DESCRIPTION
**Pull request by j4k4**

### CODE REVIEW

File: pkg/sns/topic.go [hunk 1]
```diff
@@ -182,3 +212,66 @@ func (t *snsTopic) listSubscriptions() ([]*sns.Subscription, error) {
...
+
+	if err = json.Unmarshal([]byte(*actualFilterPolicy), &actualAttributes); err != nil {
+		return false, fmt.Errorf("can not unmarshal actual filter policy: %w", err)
+	}
+
+	if expectedFilterPolicy, err = json.Marshal(attributes); err != nil {
+		return false, fmt.Errorf("can not marshal expected filter policy: %w", err)
+	}
+
+	if err = json.Unmarshal(expectedFilterPolicy, &expectedAttributes); err != nil {
```
Comment (ajscholl) [inline pkg/sns/topic.go:240]: Let me guess, this converts numbers in the map to float64 and similar?

File: pkg/sns/topic_test.go [hunk 2]
```diff
@@ -4,36 +4,29 @@ import (
...
-
-	exec := gosoAws.NewTestableExecutor([]gosoAws.TestExecution{{
-		Output: nil,
-		Err:    nil,
-	}})
+type TopicTestSuite struct {
+	suite.Suite
+	client   *snsMocks.Client
+	executor *gosoAws.TestableExecutor
+	topic    *sns.SnsTopic
```
Comment (ajscholl) [inline pkg/sns/topic_test.go:22]: Can this be an interface, so SnsTopic does not need to be public?

File: pkg/sns/topic_test.go [hunk 3]
```diff
@@ -43,133 +36,179 @@ func TestTopic_Publish(t *testing.T) {
...
-	exec := gosoAws.NewTestableExecutor([]gosoAws.TestExecution{{
-		Output: nil,
-		Err:    errors.New("error"),
-	}})
+	err := s.topic.Publish(context.Background(), aws.String("test"), map[string]interface{}{})
+	s.NoError(err)
 
-	client := new(snsMocks.Client)
-	client.On("PublishRequest", input).Return(nil, nil)
+	s.True(s.executor.HasExecutedAll())
```
Comment (ajscholl) [inline pkg/sns/topic_test.go:63]: If this is the only use of `HasExecutedAll`, maybe it should be replaced by some `AssertExpectations` function

File: pkg/sns/topic.go [hunk 4]
```diff
@@ -26,14 +29,14 @@ type Settings struct {
 	Backoff exec.BackoffSettings
 }
 
-type snsTopic struct {
+type SnsTopic struct {
```
Comment (ajscholl) [inline pkg/sns/topic.go:32]: We could leave this private if we add another interface which also includes `SubscribeSqs` (would be enough to have this in the test only)

File: pkg/sns/topic.go (hunk 1 above)
Comment (j4k4) [reply pkg/sns/topic.go:240]: yes

File: pkg/sns/topic_test.go (hunk 3 above)
Comment (j4k4) [reply pkg/sns/topic_test.go:63]: yeah, that was my first idea, too, but then you have a dependency to the testing package in `cloud/aws`..  
that doesn't felt nice..

File: pkg/sns/topic.go (hunk 4 above)
Comment (j4k4) [reply pkg/sns/topic.go]: okay

File: pkg/sns/topic_test.go (hunk 2 above)
Comment (j4k4) [reply pkg/sns/topic_test.go]: see below

File: pkg/sns/topic.go (hunk 1 above)
Comment (selm0) [reply pkg/sns/topic.go:241]: Would be nice to have an easier way to do that :thinking: Does mapstruct allow serialization from struct to struct? I know it allows weak typing

File: pkg/sns/topic_test.go (hunk 3 above)
Comment (selm0) [reply pkg/sns/topic_test.go:63]: Does it make sense to move the TestableExecutor to a test package? As it's only supposed to be used in tests? There you would be able to depend on the testing package, right?

Or replace the executor entirely by a mock?
I don't see the benefit of having a stub here, especially as you're reimplementing the expectations

**Review by ajscholl — APPROVED** [review]

**Pull request merged**